include/pmacc/algorithms/ForEach.hpp

```diff
@@ -0,0 +1,210 @@
+/* Copyright 2019 Sergei Bastrakov
+*
+* This file is part of PMacc.
+*
+* PMacc is free software: you can redistribute it and/or modify
+* it under the terms of either the GNU General Public License or
+* the GNU Lesser General Public License as published by
+* the Free Software Foundation, either version 3 of the License, or
+* (at your option) any later version.
+*
+* PMacc is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License and the GNU Lesser General Public License
+* for more details.
+*
+* You should have received a copy of the GNU General Public License
+* and the GNU Lesser General Public License along with PMacc.
+* If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include "pmacc/traits/GetNumWorkers.hpp"
+#include "pmacc/types.hpp"
+
+#include <cstdint>
+
+
+namespace pmacc
+{
+namespace fields
+{
+namespace algorithm
+{
+namespace acc
+{
+namespace detail
+{
+
+    /// TODO
+    template<
+        uint32_t T_numWorkers,
+        uint32_t T_xChunkSize
+    >
+    struct ForEachFieldValue
+    {
+        /// TODO
+        template<
+            typename T_Acc,
+            typename T_DataBox,
+            typename T_Size
+        >
+        DINLINE void operator( )(
+            T_Acc const & acc,
+            T_DataBox box,
+            T_Size size
+        ) const
+        {
+            using namespace mappings::threads;
+
+            using SuperCellSize = typename T_Mapping::SuperCellSize;
+            constexpr uint32_t dim = SuperCellSize::dim;
+            using namespace mappings::threads;
+            constexpr uint32_t cellsPerSuperCell = pmacc::math::CT::volume< SuperCellSize >::type::value;
+            uint32_t const workerIdx = threadIdx.x;
+            DataSpace< dim > const superCellIdx =
+                mapper.getSuperCellIndex( DataSpace< dim >( blockIdx ) );
+            DataSpace< dim > const blockCell = superCellIdx * SuperCellSize::toRT( );
+
+            DataSpace< dim > const localSuperCellOffset =
+                superCellIdx - mapper.getGuardingSuperCells( );
+            auto accFunctor = functor(
+                acc,
+                localSuperCellOffset,
+                WorkerCfg< T_numWorkers >{ workerIdx }
+            );
+
+            auto const blockIndex{ blockIdx };
```

Comment:
please use `DataSpace< dim >( blockIdx )` because `blockIdx` is an CUDA build in type and lags dimension information.

Reply:
NOTE: `blockIndex` will not give you any real position within the simulation domain. `superCellIdx` is the index to a supercell.